File: Spiral/GameKitHelper.m
//
//  GameKitHelper.m
//  ColorAtom
//
//  Created by 杨萧玉 on 14-9-12.
//  Copyright (c) 2014年 杨萧玉. All rights reserved.
//

#import "GameKitHelper.h"
#import <EXTScope.h>
#import "Spiral-swift.h"


@interface GameKitHelper () <GKGameCenterControllerDelegate> {
    BOOL _gameCenterFeaturesEnabled;
}
@end

@implementation GameKitHelper
@synthesize achievementsDictionary;

#pragma mark Singleton stuff
+(id) sharedGameKitHelper {
    static GameKitHelper *sharedGameKitHelper;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedGameKitHelper =
        [[GameKitHelper alloc] init];
    });
    return sharedGameKitHelper;
}

-(instancetype) init{
    if (self = [super init]) {
        achievementsDictionary = [[NSMutableDictionary alloc] init];
    }
    return self;
}
#pragma mark Player Authentication

-(void) authenticateLocalPlayer {
    
    GKLocalPlayer* localPlayer =
    [GKLocalPlayer localPlayer];
    @weakify(localPlayer)
    localPlayer.authenticateHandler =
    ^(UIViewController *viewController,
      NSError *error) {
        @strongify(localPlayer)
        [self setLastError:error];
        if (localPlayer.authenticated) {
            _gameCenterFeaturesEnabled = YES;
            if (_submitScoreWithCompletionHandler) {
                _submitScoreWithCompletionHandler(YES);
            }
            [self loadAchievements];
        } else if(viewController) {
            //TODO:palse
            [self pause];
            [self presentViewController:viewController];
        } else {
            _gameCenterFeaturesEnabled = NO;
            if (_submitScoreWithCompletionHandler) {
                _submitScoreWithCompletionHandler(NO);
            }
        }
    };
}

#pragma mark Property setters

-(void) setLastError:(NSError*)error {
    _lastError = [error copy];
    if (_lastError) {
        NSLog(@"GameKitHelper ERROR: %@", [[_lastError userInfo]
                                           description]);
    }
}

#pragma mark UIViewController stuff

-(UIViewController*) getRootViewController {
    return [UIApplication
            sharedApplication].keyWindow.rootViewController;
}

-(void)presentViewController:(UIViewController*)vc {
    UIViewController* rootVC = [self getRootViewController];
    [rootVC presentViewController:vc animated:YES
                       completion:nil];
}

#pragma mark GameKitHelperProtocol
-(void) submitScore:(int64_t)score
           identifier:(NSString*)identifier {
    //1: Check if Game Center
    //   features are enabled
    if (!_gameCenterFeaturesEnabled) {
        //@"Player not authenticated"
        if (_submitScoreWithCompletionHandler) {
            _submitScoreWithCompletionHandler(NO);
        }
        return;
    }
    
    //2: Create a GKScore object
    GKScore* gkScore = [[GKScore alloc] initWithLeaderboardIdentifier:identifier];
    //3: Set the score value
    gkScore.value = score;
    
    //4: Send the score to Game Center
    [GKScore reportScores:@[gkScore] withCompletionHandler:
     ^(NSError* error) {
         
         [self setLastError:error];
         
         BOOL success = (error == nil);
         if (_submitScoreWithCompletionHandler) {
             _submitScoreWithCompletionHandler(success);
         }
         if ([_delegate
              respondsToSelector:
              @selector(onScoresSubmitted:)]) {
             
             [_delegate onScoresSubmitted:success];
         }
     }];
}

-(void)pause{
    [[NSNotificationCenter defaultCenter] postNotificationName:WantGamePauseNotification object:nil];
//    GameViewController* gvc = (GameViewController*)[self getRootViewController];
//    [((GameScene*)(((SKView*)gvc.view).scene)) pause];
}

#pragma mark Achievements Methods
- (void) loadAchievements
{
    [GKAchievement loadAchievementsWithCompletionHandler:^(NSArray *achievements, NSError *error)
     {
         if (error == nil)
         {
             for (GKAchievement* achievement in achievements)
                 [achievementsDictionary setObject: achievement forKey: achievement.identifier];
         }
     }];
}

- (GKAchievement*) getAchievementForIdentifier: (NSString*) identifier
{
    GKAchievement *achievement = [achievementsDictionary objectForKey:identifier];
    if (achievement == nil)
    {
        achievement = [[GKAchievement alloc] initWithIdentifier:identifier];
        [achievementsDictionary setObject:achievement forKey:achievement.identifier];
    }
    return achievement;
}

- (void) updateAchievement:(GKAchievement*) achievement Identifier: (NSString*) identifier
{
    if (achievement)
    {
        [self.achievementsDictionary setObject:achievement forKey:identifier];
    }
}

- (void) reportMultipleAchievements
{
    
    [GKAchievement reportAchievements:[self.achievementsDictionary allValues]  withCompletionHandler:^(NSError *error)
     {
         if (error != nil)
         {
             NSLog(@"Error in reporting achievements: %@", error);
         }
     }];
}

- (void) showLeaderboard
{
    GKGameCenterViewController *gameCenterController = [[GKGameCenterViewController alloc] init];
    if (gameCenterController != nil)
    {
        [self pause];
        gameCenterController.gameCenterDelegate = self;
        gameCenterController.viewState = GKGameCenterViewControllerStateAchievements;
        [self presentViewController: gameCenterController];
    }
}

#pragma mark GKGameCenterControllerDelegate
- (void)gameCenterViewControllerDidFinish:(GKGameCenterViewController *)gameCenterViewController
{
    [[self getRootViewController] dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Spiral/GameKitHelper.swift
//
//  GameKitHelper.swift
//  Spiral
//
//  Created by 杨萧玉 on 15/9/13.
//  Copyright © 2015年 杨萧玉. All rights reserved.
//

import Foundation
import GameKit

let kHighScoreLeaderboardIdentifier = "com.yulingtianxia.Spiral.HighScores"
let kClean100KillerAchievementID = "com.yulingtianxia.Spiral.clean100Killer"
let kCatch500ScoreAchievementID = "com.yulingtianxia.Spiral.catch500Score"
let kCatch500ShieldAchievementID = "com.yulingtianxia.Spiral.catch500Shield"
let kget50PointsAchievementID = "com.yulingtianxia.Spiral.get50"
let kget100PointsAchievementID = "com.yulingtianxia.Spiral.get100"
let kget200PointsAchievementID = "com.yulingtianxia.Spiral.get200"


let WantGamePauseNotification = "gamepause"


protocol GameKitHelperProtocol: NSObjectProtocol {
    func onScoresSubmitted(_ success:Bool)
}

class GameKitHelper: NSObject, GKGameCenterControllerDelegate {
    var delegate : GameKitHelperProtocol?
    var lastError : NSError? {
        willSet(newError) {
            if newError != nil {
                print("GameKitHelper ERROR: \(newError!.userInfo.description)")
            }
        }
    }
    var achievementsDictionary: [String:GKAchievement]
    var submitScoreWithCompletionHandler: ((_ success: Bool)->Void)?
    
    fileprivate var gameCenterFeaturesEnabled: Bool = false
    
    fileprivate static let sharedInstance = GameKitHelper()
    
    class var sharedGameKitHelper: GameKitHelper {
        return sharedInstance
    }
    
    override init() {
        achievementsDictionary = [String:GKAchievement]()
        super.init()
        
    }
    
    //MARK: - Player Authentication
    
    func authenticateLocalPlayer() {
        let localPlayer = GKLocalPlayer.localPlayer()
        localPlayer.authenticateHandler = { [unowned localPlayer] (viewController, error) in
            self.lastError = error as NSError?
            if localPlayer.isAuthenticated {
                self.gameCenterFeaturesEnabled = true
                self.submitScoreWithCompletionHandler?(true)
                self.loadAchievements()
            }
            else if let vc = viewController {
                self.pause()
                self.presentViewController(vc)
            }
            else {
                self.gameCenterFeaturesEnabled = false
                self.submitScoreWithCompletionHandler?(false)
            }
        }
    }
    
    //MARK: - UIViewController stuff
    
    func getRootViewController() -> UIViewController? {
        return UIApplication.shared.keyWindow?.rootViewController;
    }
    
    func presentViewController(_ vc: UIViewController) {
        let rootVC = self.getRootViewController()
        rootVC?.present(vc, animated: true, completion: nil)
    }
    
    // MARK: - GameKitHelperProtocol
    
    func submitScore(_ score: Int64, identifier: String) {
        //1: Check if Game Center features are enabled
        guard gameCenterFeaturesEnabled else {
            submitScoreWithCompletionHandler?(false)
            return
        }
        //2: Create a GKScore object
        let gkScore = GKScore(leaderboardIdentifier: identifier)
        //3: Set the score value
        gkScore.value = score
        //4: Send the score to Game Center
        GKScore.report([gkScore], withCompletionHandler: { (error) -> Void in
            self.lastError = error as NSError?
            let success = (error == nil)
            self.submitScoreWithCompletionHandler?(success)
            self.delegate?.onScoresSubmitted(success)
        }) 
    }
    
    func pause() {
        NotificationCenter.default.post(name: Notification.Name(rawValue: WantGamePauseNotification), object: nil)
    }
    
    //MARK: - Achievements Methods
    
    func loadAchievements() {
        GKAchievement.loadAchievements { (achievements, error) -> Void in
            if let achievements = achievements , error == nil {
                for achievement in achievements {
                    self.achievementsDictionary[achievement.identifier!] = achievement
                }
            }
        }
    }
    
    func getAchievementForIdentifier(_ identifier: String) -> GKAchievement {
        guard let achievement = achievementsDictionary[identifier] else {
            let achievement = GKAchievement(identifier: identifier)
            achievementsDictionary[achievement.identifier!] = achievement
            return achievement;
        }
        return achievement;
    }
    
    func updateAchievement(_ achievement: GKAchievement, identifier: String) {
        achievementsDictionary[identifier] = achievement
    }
    
    func reportMultipleAchievements() {
        GKAchievement.report(Array(achievementsDictionary.values), withCompletionHandler: { (error) -> Void in
            if error != nil {
                fatalError("Error in reporting achievements:\(error)")
            }
        }) 
    }
    
    func showLeaderboard() {
        let gameCenterController = GKGameCenterViewController()
        pause()
        gameCenterController.gameCenterDelegate = self
        gameCenterController.viewState = .achievements
        presentViewController(gameCenterController)
    }
    
    //MARK: - GKGameCenterControllerDelegate
    func gameCenterViewControllerDidFinish(_ gameCenterViewController: GKGameCenterViewController) {
        getRootViewController()?.dismiss(animated: true, completion: nil)
    }
}
